src/storage/mvcc/reader/scanner/forward.rs

```diff
@@ -1240,3 +1433,533 @@ mod latest_entry_tests {
+    #[test]
+    fn test_move_next_user_key_out_of_bound_2() {
+        let engine = TestEngineBuilder::new().build().unwrap();
+
+        // Generate 1 put for [a].
+        must_prewrite_put(&engine, b"a", b"a_value", b"a", SEEK_BOUND * 2);
+        must_commit(&engine, b"a", SEEK_BOUND * 2, SEEK_BOUND * 2);
+
+        // Generate SEEK_BOUND rollback and 1 put for [b] .
+        // It differs from EntryScanner that this will try to fetch multiple versions of each key.
+        // So in this test it needs one more next than EntryScanner.
+        for ts in 1..=SEEK_BOUND {
+            must_rollback(&engine, b"b", ts as u64);
+        }
+        must_prewrite_put(&engine, b"b", b"b_value", b"a", SEEK_BOUND + 1);
+        must_commit(&engine, b"b", SEEK_BOUND + 1, SEEK_BOUND + 1);
+
+        let snapshot = engine.snapshot(&Context::default()).unwrap();
+        let mut scanner = ScannerBuilder::new(snapshot, (SEEK_BOUND * 2).into(), false)
+            .range(None, None)
+            .build_delta_scanner(8.into())
+            .unwrap();
+
+        // The following illustration comments assume that SEEK_BOUND = 4.
+
+        // Initial position: 1 seek_to_first:
+        //   a_8 b_4 b_3 b_2 b_1
+        //   ^cursor
+        // After get the value, use 1 next to reach next user key:
+        //   a_8 b_4 b_3 b_2 b_1
+        //       ^cursor
+        let entry = EntryBuilder::default()
+            .key(b"a")
+            .value(b"a_value")
+            .start_ts(16.into())
+            .commit_ts(16.into())
+            .build_commit(WriteType::Put, true);
+        assert_eq!(scanner.next_entry().unwrap(), Some(entry));
+        let statistics = scanner.take_statistics();
+        assert_eq!(statistics.write.seek, 1);
+        assert_eq!(statistics.write.next, 1);
+
+        // Before:
+        //   a_8 b_4 b_3 b_2 b_1
+        //       ^cursor
+        // We should be able to get wanted value without any operation.
+        // After get the value, use SEEK_BOUND-1 next: (TODO: fix it to SEEK_BOUND)
+        //   a_8 b_4 b_3 b_2 b_1
+        //                   ^cursor
+        // We still pointing at current user key, so a seek:
+        //   a_8 b_4 b_3 b_2 b_1
+        //                       ^cursor
+        let entry = EntryBuilder::default()
+            .key(b"b")
+            .value(b"b_value")
+            .start_ts(9.into())
+            .commit_ts(9.into())
+            .build_commit(WriteType::Put, true);
+        assert_eq!(scanner.next_entry().unwrap(), Some(entry),);
+        let statistics = scanner.take_statistics();
+        assert_eq!(statistics.write.seek, 0);
+        assert_eq!(statistics.write.next, 1);
+
+        // Next we should get nothing.
+        assert_eq!(scanner.next_entry().unwrap(), None);
+        let statistics = scanner.take_statistics();
+        assert_eq!(statistics.write.seek, 1);
+        assert_eq!(statistics.write.next, (SEEK_BOUND - 1) as usize);
+    }
+
+    /// Range is left open right closed.
```

Comment:
Could you specify key range is left close right open and time range is left open right close?

Reply:
This test case didn't check this.  It's checked in `test_mess`.